### ISSUE TITLE
hyprctl switchxkblayout lagging 

### ISSUE THREAD

**Issue by coffebar** [hyprwm/Hyprland#7044]
### Regression?

Yes

### System Info and Version

<details> <summary>System/Version info</summary>

```sh
Hyprland, built from branch main at commit 99088eaed806d9268967baf09bc09cdb987c5357  (compositor: simplify getWindowFromSurface).
Date: Wed Jul 24 10:07:28 2024
Tag: v0.41.2-84-g99088eae, commits: 4970

flags: (if any)


System Information:
System name: Linux
Node name: ***
Release: 6.10.0-arch1-2
Version: #1 SMP PREEMPT_DYNAMIC Mon, 22 Jul 2024 17:28:23 +0000


GPU information:
00:02.0 VGA compatible controller [0300]: Intel Corporation TigerLake-LP GT2 [Iris Xe Graphics] [8086:9a49] (rev 01) (prog-if 00 [VGA controller])


os-release: NAME="Arch Linux"
```

</details>

### Description

Here is the video where layouts don't match what hyprctl shows and what I see when type. 
This bug seems to appear randomly 

### How to reproduce

I use `hyprctl switchxkblayout ACTIVE_KEYBOARD_NAME INDEX` to switch layouts
and 
`hyprctl devices -j | jq '.keyboards.[] | select(.main == true).active_keymap'` to see active keymap, 
I'm typing to see actual keymap (Latin or Cyrillic) 
Index 0 - us_en
Index 1 - uk Cyrillic

`watch` command is helpful to automate getting results 

```
# hyprland.conf
input {
  kb_layout = us,ua
  kb_variant =
  kb_model =
  kb_options = grp:alt_shift_toggle,caps:backspace
  kb_rules =
}
```

### Crash reports, logs, images, videos

https://github.com/user-attachments/assets/3dbc833a-71ee-42cd-a659-2b65e57aede3

`watch "hyprctl devices -j | jq '.keyboards.[] | select(.main == true).active_keymap'"`

`hyprctl switchxkblayout keychron-keychron-k2 1`

**Comment by vaxerski** [comment]
can you bisect if this is a regression

**Comment by coffebar** [comment]
I rolled back to v0.41.1, and it is not affected. 
v0.41.2 is affected for sure

**Comment by MahouShoujoMivutilde** [comment]
You're using default `watch` refresh rate which is every 2s. In fact, `watch` can't go faster than every 100ms. I don't know why it works for you on 0.41.1 better.

Trying to recreate the same demo on e728e56cbc6af0a9be1276b2cf8e019a894016e7 (0.41.2 pre aquamarine):


https://github.com/user-attachments/assets/121de973-7582-49de-9694-f9b7ed439d88

Seems to be working fine.

```bash
while true; do
    hyprctl devices -j | jq '.keyboards.[] | select(.main == true).active_keymap';
    sleep 0.01s; # because watch sucks
done

```

Did aquamarine break something? 

Here is 33a5c8ce325e1657ec5571b57e05c6378706f4cd with latest aquamarine/hyprutils/hyprlang (disregard the comment on the right, it was recorded out of order):


https://github.com/user-attachments/assets/3f02aad0-4877-494c-bded-cd8e261a8d14

And here I set russian layout for the left window and english for the right, via _my_ [hyprland-per-window-layout](https://github.com/MahouShoujoMivutilde/hyprland-per-window-layout).

https://github.com/user-attachments/assets/d8cb7e20-bc90-4971-a749-18a659516da8

I was trying to input `w` and `ц` (which are the same physical key) as fast as I could to show that layout actually changes, and that hyprctl reflects it correctly. And then I was switching as fast as possible to try to break it.

Seems to work fine?

But notice how waybar's layout is broken.

**Comment by coffebar** [comment]
You are trying to do it differently, I'm using Indexes when switching. Also your config is different. I will try to experiment with this. 

2 sec is fast enough for video to see that the layout was changed, I don't make an effort to break it with fast moves

**Comment by coffebar** [comment]
@MahouShoujoMivutilde I tried your way and your config on v0.41.2-84-g99088eae


```
device {
    name = keychron-keychron-k2
    kb_layout = us,ru
    kb_options = caps:escape,grp:alt_shift_toggle
}
```

and it reproduced as well:

https://github.com/user-attachments/assets/0769b3a6-12ac-4e25-ab32-c1528920a08f

`watch -n 0.5 "hyprctl devices -j | jq '.keyboards.[] | select(.main == true).active_keymap'"`

When I'm using Alt+Shift it actually changes the layout well, but with `hyprctl switchxkblayout keychron-keychron-k2 next` - some mess

**Comment by izmyname** [comment]
Looks like I have a somewhat similar or related problem.

`"hyprctl switchxkblayout at-translated-set-2-keyboard next`
doesn't switch the keyboard, although

 ```
input {
   kb_layout = us,ru
   kb_variant =
   kb_model =
   kb_options = grp:alt_shift_toggle
   kb_rules =
```
in my config allows me to switch the layouts with shift+alt key combination.

**Comment by MahouShoujoMivutilde** [comment]
@coffebar sorry, it was late, and I didn't notice that you checked with typing too.

Tested it some more. This is on aquamarine 0.1.0_r157.g0ab8ffa-1 and Hyprland 33a5c8ce325e1657ec5571b57e05c6378706f4cd

When layout is defined in general:
* switching via hotkey - events are emitted, active keymap in `hyprctl devices -j` is updated, and the right letters **are used**.
* via `hyprctl switchxkblayout` - events are **not** emitted, active keymap is updated, and right letters **are not** used (last layout switched by hotkey used).


When layout is defined per device:
* layout can be switched by hotkey - events are emitted, active keymap in `hyprctl devices -j` is updated, **and right letters are used**.
* via `hyprctl switchxkblayout` - events are **not** emitted, active keymap is updated, and right letters **are not** used (last layout switched by hotkey used).

WTF? It worked just this night. More so, it worked when I first booted just now.

But, for both general and per device:

* When `switchxkblayout` doesn't _actually_ switch - you can **press Shift** (and nothing else, it's not a `grp:` hotkey), and **then** events will be emitted, and layout will change for real.

I kid you not, I discovered this by trying to write 'WTF' to mark the moment of it glitching out in recording.

Okay, so something fucky with the state is going on.

I'll try to reboot and recreate my original demos, see at what point they break.

---

**You're right, it's broken.** So why did it work for me earlier?

Modifier keys! Super | Ctrl | Shift | Alt.

I was holding Super to switch between windows. It forced proper layout updates like the Shift did, above.

In the first comment, you say that its "random".

My guess is that you're experiencing this issue with your layout switcher, and you too switch windows using one of the mod keys, and sometimes you don't hold it long enough for it to manage to switch layout for real.

So it **is** a problem that didn't happen before, it's just that it can be weirdly worked around by holding mod keys longer, so I didn't notice it initially.

---

Here I switch layout via hotkey a few times, type to show that it works, and then use hyprctl to switch to russian, type to check, it didn't work, and then I press Super and it changes layout for real.


https://github.com/user-attachments/assets/2e2db561-3618-43b8-814d-bd18618916dd

**Comment by MahouShoujoMivutilde** [comment]
Alright, I've tried to recreate what's going on when you press mod keys, so here is the patch that """fixes""" the issue.

<details><summary>Patch</summary>
<p>

```diff
diff --git a/src/devices/Keyboard.cpp b/src/devices/Keyboard.cpp
index 0a8f6b57..48159339 100644
--- a/src/devices/Keyboard.cpp
+++ b/src/devices/Keyboard.cpp
@@ -38,6 +38,15 @@ CKeyboard::CKeyboard(SP<Aquamarine::IKeyboard> keeb) : keyboard(keeb) {
             .state   = E.pressed ? WL_KEYBOARD_KEY_STATE_PRESSED : WL_KEYBOARD_KEY_STATE_RELEASED,
         });

+        // updateModifiersState(); // not enough
+
+        keyboardEvents.modifiers.emit(SModifiersEvent{
+            .depressed = modifiersState.depressed,
+            .latched   = modifiersState.latched,
+            .locked    = modifiersState.locked,
+            .group     = modifiersState.group,
+        });
+
         updateXkbStateWithKey(E.key + 8, E.pressed);
     });

```


</p>
</details> 

It probably should be more concrete, like seeing what actually unfucks the layout during the processing of modifiers and doing that. At least it doesn't _seem_ to break anything.

I would investigate further, but I don't really have an environment for C++ set up to do all this "go to references" etc, this is really just courtesy of `ripgrep`.

---

The patch in action:

Defined per device


https://github.com/user-attachments/assets/10a86769-c53e-49a1-908e-2ca1360739c7



Defined in global input


https://github.com/user-attachments/assets/4c16dcbb-18ab-4c64-9bb2-b8a7c582e63d


Now, witch the patch, on `hyprctl` layout switching:
* layout is really changed
* event is emitted, as it should
* active keymap updated in `hyprctl devices`

**Comment by MahouShoujoMivutilde** [comment]
Never mind, found a better place to put it, here is the patch:

<details><summary>Details</summary>
<p>

```diff
diff --git a/src/devices/IKeyboard.cpp b/src/devices/IKeyboard.cpp
index 09a18477..e3517bf1 100644
--- a/src/devices/IKeyboard.cpp
+++ b/src/devices/IKeyboard.cpp
@@ -325,6 +325,13 @@ void IKeyboard::updateModifiers(uint32_t depressed, uint32_t latched, uint32_t l
     if (!updateModifiersState())
         return;
 
+    keyboardEvents.modifiers.emit(SModifiersEvent{
+        .depressed = modifiersState.depressed,
+        .latched   = modifiersState.latched,
+        .locked    = modifiersState.locked,
+        .group     = modifiersState.group,
+    });
+
     updateLEDs();
 }
 

```

</p>
</details> 

Now, with the patch, on hyprctl layout switching:
* layout is really changed
* event is emitted, as it should
* active keymap updated in hyprctl devices

https://github.com/user-attachments/assets/9d74701b-e993-4233-9d05-8870b0cb5012

(`sleep 1s` so I can see that it changes because of hyprctl request, and nothing else)

Seems to just work at this point?

**Comment by coffebar** [comment]
Applied this patch, seems to work. Did you figure out what happened to broke it?

**Comment by MahouShoujoMivutilde** [comment]
It seems earliest version of `IKeyboard::updateModifiers()`

https://github.com/hyprwm/Hyprland/blame/7dee974be8915864e05e1b3cc8bb3eb3fdb7fc90/src/devices/IKeyboard.cpp#L290-L300

didn't ever emit event (which, apparently, fixes layout somewhere down the line).

This function, at least without some extra effort to find its earlier variations unseen by git blame, only appears with aquamarine.

So, migration to aquamarine broke it. 

I didn't bisect for this exact reason - suspected it'll come down to that anyway.

**Comment by coffebar** [comment]
Something is broken here after the update v0.42.0-73-g1ea47950 AUR
When I'm using hyprland-per-window-layout - my layouts go crazy

**Comment by MahouShoujoMivutilde** [comment]
What's the version you upgraded from, that worked?

I'm on v0.42.0-65-g66586c38 and `hyprctl switchxkblayout` works fine, my script works, your app works too.

---

After some debugging I can say that it's related to config autoreload.

`switchxkblayout` seems to be fine. Is maintaining the correct hashtable mapping windows to layouts right after config reload the issue here?

Recently I noticed that focused window changes after reload, if you had multiple windows opened - it always focuses window under the cursor for me (just checked with floating windows to make sure); mouse is also centered after reload.

Maybe what's happening is that right after reload events are messed up...

I also did https://github.com/hyprwm/Hyprland/pull/7486 that influences mod state after reload (if `numlock_by_default` is on) and thus emits the same (internal) event #7070 added, but that shouldn't matter, I think.

Let's see.

<details><summary>Details</summary>
<p>

```
activewindow>>foot,~ - zsh
activewindowv2>>562804a873e0
activelayout>>logitech-usb-keyboard,Russian
activelayoutv2>>logitech-usb-keyboard\,Russian
# ❗Switched to russian, and running hyprctl reload in 562804a873e0 window,
#  which is at the side, with 5628057fea20 at the center
activelayout>>power-button,English (US)
activelayoutv2>>power-button\,English (US)
activelayout>>power-button-1,English (US)
activelayoutv2>>power-button-1\,English (US)
activelayout>>logitech-usb-keyboard,English (US)           # ❗lost right layout for this window here
activelayoutv2>>logitech-usb-keyboard\,English (US)
activelayout>>logitech-usb-keyboard-consumer-control,English (US)
activelayoutv2>>logitech-usb-keyboard-consumer-control\,English (US)
activelayout>>logitech-usb-keyboard-system-control,English (US)
activelayoutv2>>logitech-usb-keyboard-system-control\,English (US)
activelayout>>eee-pc-wmi-hotkeys,English (US)
activelayoutv2>>eee-pc-wmi-hotkeys\,English (US)
configreloaded>>
activewindow>>foot,~/projects/installed/coffebar-hyprland-per-window-layout/target/release - zsh
activewindowv2>>5628057fea20
activelayout>>logitech-usb-keyboard,English (US)   # ❗❗❗why again, and after automatic focus change at that?
activelayoutv2>>logitech-usb-keyboard\,English (US)ac
```

</p>
</details>

The layout is **still** not preserved when `numlock_by_default = no` ... because I forgot about it in stuck mods MR.

But when it is `yes` it was preserved **anyway** (before my MRs), so probably not the cause here - overall, layout-related behavior there didn't change.

<details><summary>Details</summary>
<p>

```
activewindow>>foot,~ - zsh
activewindowv2>>562804a873e0
activelayout>>logitech-usb-keyboard,Russian
activelayoutv2>>logitech-usb-keyboard\,Russian
# ❗Switched to russian, and running hyprctl reload in 562804a873e0 window (at the side), 5628057fea20 (center)
activelayout>>power-button,English (US)
activelayoutv2>>power-button\,English (US)
activelayout>>power-button-1,English (US)
activelayoutv2>>power-button-1\,English (US)
activelayout>>logitech-usb-keyboard,Russian
activelayoutv2>>logitech-usb-keyboard\,Russian        # ❗right layout for focused window is preserved... 
#                                                                                             but focus is already changed
activelayout>>logitech-usb-keyboard-consumer-control,English (US)
activelayoutv2>>logitech-usb-keyboard-consumer-control\,English (US)
activelayout>>logitech-usb-keyboard-system-control,English (US)
activelayoutv2>>logitech-usb-keyboard-system-control\,English (US)
activelayout>>eee-pc-wmi-hotkeys,English (US)
activelayoutv2>>eee-pc-wmi-hotkeys\,English (US)
configreloaded>>
activewindow>>foot,~/projects/installed/coffebar-hyprland-per-window-layout/target/release - zsh
activewindowv2>>5628057fea20                                 # and now it actually noticed that focus changed
# ❗no extra activelayout events
```

</p>
</details>


Even if layout was always preserved (not only when `numlock_by_default = yes`) on reload, it'll still be a problem when you have:
* focused window with layout X
* another that will be under cursor (should it be centered) with saved layout Y

Layout X will redefine layout Y for this other window on reload.

But, if layout is not preserved - it'll be redefined even if it's a single window on workspace.

But focus shouldn't change at all on reload. I don't yet understand why it does.

**It all comes down to it.**


---

Hold up.

Basic stuff.

Sometimes, somehow `switchxkblayout` doesn't actually change the layout right after reload when its inside (your) hyprland-per-window-layout. I triggered it once, and can't reproduce anymore. Idk.

Is this is what you mean? Can you give me a way to reproduce it?

**Comment by coffebar** [comment]
My app runs `hyprctl devices -j` on startup and see this: 
```json
{
    "mice": [

    ],
    "keyboards": [

    ],
    "tablets": [

    ],
    "touch": [

    ],
    "switches": [

    ]
}
```

However, when I run this command later, it shows actual devices.

**Comment by MahouShoujoMivutilde** [comment]
Ah, that.

That's an unfortunate change that I noticed too, but I don't know what caused it.

I didn't bother investigating it, I just added the back-off strategy on my script's init - `devices` are queried up to 3 times, until the keyboard is found, with progressively longer `sleep` at each failure.

**Comment by MahouShoujoMivutilde** [comment]
Also noticed that `main` in `devices` is unreliable at startup.

If layout configured in general `input`, at hyprland's launch at first there is nothing in devices, then `eee-pc-wmi-hotkeys` (for me) has `main: true`, and only then my actual keyboard.

**Comment by slevi123** [comment]
It happened to me too.

I created a bash script, bound to SUPER+space in `hyprland.conf`.
It used `hyprctl switchkbd DEVICE_NAME LAYOUT_NUMBER` to switch layout. When I run the script from terminal it worked, but when I used the shortcut, it was only reflected on waybar, the keys did not work.

I put `sleep 0.1` into my script and it resolved it.
Thank you!

Some aditional info: 
I am on NixOS 24.11.20250126.4e96537.

My hyperland version:
![Image](https://github.com/user-attachments/assets/8bc4b555-b2f6-4aeb-bad1-202cd588bbf8)
